Signature: Make key management Actor-specific

obenland opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

Minor house keeping item in preparation for #1808, separating the key management bits out from the signature functionality itself. I'm hoping it'll make it easier to update the Signature class in the future.

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Moves key management functions to `Actors` class.
* Updates all references to those functions.
* Deprecates public functions in Signature class.
* Moves unit tests.

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Subscribe to your test site from the Fediverse.
* Publish a new post.
* Make sure it still gets delivered to your subscriber and signature management still works.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger a GitHub workflow that will create and push the entry into the branch. -->

<!-- Due to org permissions, the job may fail for PRs crated from a fork under GitHub organizations. -->
<!-- In this case, you can create entry manually with `composer changelog:add` and push it into the branch. -->

<!-- If no changelog entry is required for this PR, please add the "Skip Changelog" label. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Added - for new features
-   [x] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

Management of public/private keys for Actors now lives in the Actors collection, in preparation for Signature improvements down the line.

</details>

obenland committed:
Signature: Make key management Actor specific

obenland committed:
Merge branch 'trunk' into update/key-signatures

matticbot committed:
Add changelog

obenland commented:
@pfefferle Is `Actors` the best place for this?

obenland commented on includes/collection/class-actors.php:
```diff
@@ -389,4 +389,162 @@ public static function get_type_by_id( $user_id ) {
+		);
+
+		// Persist keys.
+		\add_option( $option_key, $key_pair );
```
I was actually wondering about that, too. Is there ever a time where we'd want to change the keys?

pfefferle replied:
There is a recommendation to do so, but I am not sure how well supported it is yet https://swicg.github.io/activitypub-http-signature/#key-rotation

pfefferle commented:
> Is Actors the best place for this?

I would think so!

Maybe we can unify the `get_public_key` functions a bit to allow a similar functionality for remote users?

Either in the same function:
* if `$id` is an URI, check for the public key in the `ap_actor` `post_content`
* if it is an ID, use the current function

Or a dedicated `get_remote_public_key`!?

pfefferle commented on includes/class-signature.php:
```diff
@@ -194,7 +81,7 @@ protected static function check_legacy_key_pair_for( $user_id ) {
 	 */
 	public static function generate_signature( $user_id, $http_method, $url, $date, $digest = null ) {
```
it would be awesome to pass the private key as a param instead of the `$user_id`, to be completely user agnostic.

obenland replied:
Agreed!

obenland replied:
Fixed in [2a0373c](https://github.com/Automattic/wordpress-activitypub/pull/1832/commits/2a0373cf689d6275a510547547b71038bb54fe5a). This turned out to be a bit more invasive than I hoped, since there are two data points dependent on the user.

Long term, I wonder if we should move Signature from being static to being an object that we'd instantiate with the user stuff and then call `generate_signature()` with just the request information

pfefferle replied:
I have indeed also thought about that... I think the http class would also benefit from beeing an object instead of a static class...

obenland commented:
> Maybe we can unify the get_public_key functions a bit to allow a similar functionality for remote users?

I was wondering about that, too, but couldn't think of a use case. To verify signatures on incoming requests I think we should always make remote requests to the `keyId` provided in the signature itself, even if it's from an actor in the db. I wouldn't want to rely on a cached version.

obenland committed:
Decouple Signature class from user

obenland committed:
Merge branch 'trunk' into update/key-signatures

obenland committed:
Call it what it is, actor_id

obenland committed:
Merge branch 'update/key-signatures' of https://github.com/Automattic/wordpress-activitypub into update/key-signatures

obenland committed:
Merge branch 'trunk' into update/key-signatures

obenland committed:
Fix whitespace after rebase

obenland committed:
Fix tests after 2a0373cf689d6275a510547547b71038bb54fe5a

obenland commented on includes/class-signature.php:
```diff
@@ -24,178 +24,63 @@ class Signature {
-		$user = Actors::get_by_id( $user_id );
-		$key  = self::get_private_key_for( $user->get__id() );
-
+	public static function generate_signature( $key_id, $private_key, $http_method, $url, $date, $digest = null ) {
```
Heads up that replacing `$user_id` with the required values changes the function signature.

pfefferle commented on includes/class-signature.php:
```diff
@@ -220,15 +105,13 @@ public static function generate_signature( $user_id, $http_method, $url, $date,
-
 		if ( ! empty( $digest ) ) {
-			return \sprintf( 'keyId="%s",algorithm="rsa-sha256",headers="(request-target) host date digest",signature="%s"', $key_id, $signature );
+			return \sprintf( 'keyId="%s",algorithm="rsa-sha256",headers="(request-target) host date digest",signature="%s"', $key_id . '#main-key', $signature );
```
I think the `#main-key` hash should be part of the key-id

obenland replied:
Fixed in [487f284](https://github.com/Automattic/wordpress-activitypub/pull/1832/commits/487f28436fb7af0ba31275547dc5074157297c22).

obenland committed:
Standardize actor_id to include '#main-key' in signatures

Appends '#main-key' to actor_id when generating signatures in HTTP requests and tests.

Props @pfefferle.

obenland committed:
Merge branch 'trunk' into update/key-signatures

obenland committed:
Fix errors after merge conflict resolution

obenland committed:
Merge branch 'trunk' into update/key-signatures

obenland committed:
Merge branch 'trunk' into update/key-signatures

obenland committed:
Merge branch 'trunk' into update/key-signatures

obenland committed:
Fix trunk merge disaster

obenland committed:
Backslashit

obenland committed:
Not sure why that was moved

obenland committed:
Revert unnecessary changes

obenland committed:
Revert unrelated changes

obenland committed:
Also move get_remote_key

obenland committed:
Move tests

obenland commented:
@pfefferle Third time's the charm!

pfefferle commented on includes/model/class-application.php:
```diff
@@ -219,7 +217,7 @@ public function get_public_key() {
 		return array(
 			'id'           => $this->get_id() . '#main-key',
```
shouldn't we use a function that generates the signature ID, to be sure that it is consistent?

pfefferle replied:
(but this might be a different PR)

obenland replied:
Honestly, I'm not even sure we need a "separate" signature ID from the actor ID. The hash doesn't seem to serve a purpose for as long as we only support one public key.
